linode/lke/cluster.go

```diff
@@ -351,34 +303,94 @@ func recycleLKECluster(ctx context.Context, meta *helper.ProviderMeta, id int, p
 
 // This cannot currently be handled efficiently by a DiffSuppressFunc
 // See: https://github.com/hashicorp/terraform-plugin-sdk/issues/477
-func matchPoolsWithSchema(pools []linodego.LKENodePool, declaredPools []interface{}) []linodego.LKEClusterPool {
+func matchPoolsWithSchema(pools []linodego.LKENodePool, declaredPools []interface{}) ([]linodego.LKENodePool, error) {
 	result := make([]linodego.LKENodePool, len(declaredPools))
 
-	poolMap := make(map[int]linodego.LKENodePool, len(declaredPools))
+	// Contains all unpaired pools returned by the API
+	apiPools := make(map[int]linodego.LKENodePool, len(pools))
 	for _, pool := range pools {
-		poolMap[pool.ID] = pool
+		apiPools[pool.ID] = pool
 	}
 
+	// Tracks which local pools have been processed
+	pairedDeclaredPools := make(map[int]bool)
+
+	// First let's match any pools in state with an ID
 	for i, declaredPool := range declaredPools {
-		declaredPool := declaredPool.(map[string]interface{})
+		declaredPool := declaredPool.(map[string]any)
+
+		poolID, ok := declaredPool["id"].(int)
+		if !ok {
+			return nil, fmt.Errorf("declared pool ID was not of type int")
+		}
+
+		apiPool, ok := apiPools[poolID]
+		if !ok {
+			continue
+		}
+
+		// Pair the found pool with the declared pool
+		result[i] = apiPool
+		delete(apiPools, poolID)
+		pairedDeclaredPools[i] = true
+	}
+
+	// Second, let's match pools that have all matching attributes.
+	// This is necessary because declared pools will not be populated with
+	// an ID on first apply but still have matching node pools.
+	for i, declaredPool := range declaredPools {
+		declaredPool := declaredPool.(map[string]any)
+		declaredAutoscaler := expandLinodeLKEClusterAutoscalerFromPool(declaredPool)
+
+		if _, ok := pairedDeclaredPools[i]; ok {
+			// This apiPool has already been handled in the previous step,
+			// we can skip it
+			continue
+		}
 
-		for key, pool := range poolMap {
-			if pool.Count != declaredPool["count"] || pool.Type != declaredPool["type"] {
+		for _, apiPool := range apiPools {
+			if declaredPool["type"] != apiPool.Type {
 				continue
 			}
 
-			result[i] = pool
-			delete(poolMap, key)
+			declaredCount := declaredPool["count"].(int)
+			if declaredCount == 0 {
+				if declaredAutoscaler == nil {
+					return nil, fmt.Errorf("autoscaler is null when count is 0. This is always a provider issue")
+				}
+				declaredCount = declaredAutoscaler.Min
+			}
+
+			if declaredCount != apiPool.Count {
+				continue
+			}
+
+			if (declaredAutoscaler != nil && declaredAutoscaler.Enabled) != apiPool.Autoscaler.Enabled {
+				continue
+			}
+
+			// Only compare autoscalers if the declared autoscaler is enabled
+			if declaredAutoscaler != nil && !reflect.DeepEqual(
```

Comment:
Should we also add `declaredAutoscaler.Enabled` to this if statement?

Reply:
I think that logic should be handled in the block above, this is just a second-pass to make sure there are no changes if an autoscaler block has been defined.